hoot-test/src/main/cpp/hoot/test/ProcessPool.cpp

```diff
@@ -0,0 +1,234 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2017 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "ProcessPool.h"
+
+#include <iostream>
+
+using namespace std;
+
+
+JobQueue::JobQueue()
+{
+}
+
+bool JobQueue::empty()
+{
+  _mutex.lock();
+  bool e = _jobs.empty();
+  _mutex.unlock();
+  return e;
+}
+
+bool JobQueue::contains(const QString& job)
+{
+  _mutex.lock();
+  bool c = _jobs.find(job) != _jobs.end();
+  _mutex.unlock();
+  return c;
+}
+
+QString JobQueue::pop()
+{
+  _mutex.lock();
+  QString job = *_jobs.begin();
+  _jobs.erase(_jobs.begin());
+  _mutex.unlock();
+  return job;
+}
+
+void JobQueue::push(const QString& job)
+{
+  _mutex.lock();
+  _jobs.insert(job);
+  _mutex.unlock();
+}
+
+ProcessThread::ProcessThread(bool showTestName,
+                             double waitTime,
+                             QMutex* outMutex,
+                             JobQueue* parallelJobs,
+                             JobQueue* serialJobs)
+  : _showTestName(showTestName),
+    _waitTime(waitTime),
+    _outMutex(outMutex),
+    _parallelJobs(parallelJobs),
+    _serialJobs(serialJobs),
+    _failures(0),
+    _proc(createProcess())
+{
+}
+
+int ProcessThread::getFailures()
+{
+  return _failures;
+}
+
+void ProcessThread::resetProcess()
+{
+  //  Kill the process
+  _proc->write(QString("%1\n").arg(HOOT_TEST_FINISHED).toAscii());
+  _proc->waitForFinished();
+  //  Start a new process
+  _proc.reset(createProcess());
+}
+
+QProcess* ProcessThread::createProcess()
+{
+  QProcess* proc = new QProcess();
+  proc->setProcessChannelMode(QProcess::MergedChannels);
+  QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
+  proc->setProcessEnvironment(env);
+  QString names = (_showTestName ? "--names" : "");
+  proc->start(QString("HootTest %1 --listen %2").arg(names).arg((int)_waitTime));
+  return proc;
+}
+
+void ProcessThread::run()
+{
+  if (_serialJobs != NULL)
+  {
+    processJobs(_serialJobs);
+    //  Reset the process between queues
+    resetProcess();
+  }
+  processJobs(_parallelJobs);
+
+  _proc->write(QString("%1\n").arg(HOOT_TEST_FINISHED).toAscii());
+  _proc->waitForFinished();
+}
+
+void ProcessThread::processJobs(JobQueue* queue)
+{
+  int restart_count = 0;
+  const int MAX_RESTART = 3;
+  const int READ_TIMEOUT = 3000;
+  bool working = true;
+  while (working)
+  {
+    if (!queue->empty())
```

Comment:
I think you have a race condition here. It is possible that queue->empty() could return false, then another thread pops the last job which would cause you to enter this method with a bad job name. Possibly have pop return an empty string if there are no more names, then you could pop and check for empty in a single call which would be atomic.